se/se_epub_generate_toc.py

```diff
@@ -140,28 +167,32 @@ def get_file_epub_type(soup: BeautifulSoup) -> str:
 	"""
 
 	# Try for a heading.
-	first_head = soup.find(["h1", "h2", "h3", "h4", "h5", "h6"])
+	first_head = get_first(dom, "//h1 | //h2 | //h3 | //h4 | //h5 | //h6")
+
 	if first_head is not None:
-		parent = first_head.find_parent(["section", "article"])
+		parent = get_outer_sections(first_head, ["section", "article"])
 	else:  # No heading found so go hunting for some other content.
-		paragraph = soup.find(["p", "header", "img"])  # We look for the first such item.
+		paragraph = get_first(dom, "//p | //header | //img")  # We look for the first such item.
 		if paragraph is not None:
-			parent = paragraph.find_parent(["section", "article"])
+			parent = get_outer_sections(paragraph, ["section", "article"])
 		else:
 			return ""
 
 	if parent is None:
-		parent = soup.find("body")
+		parent = dom.xpath("//body")
 
-	try:
-		return parent["epub:type"]
-	except KeyError:
+	epub_type = parent.get_attr("epub:type")
+	if not epub_type:
```

Comment:
You can test for a node that has the `epub:type` attribute using xpath. So instead of doing:

````
parent = dom.xpath("//body")
epub_type = parent.get_attr("epub:type")
if not epub_type:
    ...
````

You can do:
````
parent = dom.xpath("//body[@epub:type]") # Select any <body> elem that has an epub:type attribute